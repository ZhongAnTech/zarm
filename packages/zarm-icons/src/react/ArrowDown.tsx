import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowDown = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M494.094 733.426a41.472 41.472 0 01-20.023-11.126L140.16 388.388c-16.272-16.271-16.272-42.653 0-58.925 16.271-16.272 42.653-16.272 58.925 0L503.66 634.037 808.186 329.51c16.271-16.272 42.653-16.272 58.925 0s16.272 42.654 0 58.926L533.2 722.347c-10.607 10.607-25.508 14.3-39.106 11.08z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowDown);
export default ForwardRef;
