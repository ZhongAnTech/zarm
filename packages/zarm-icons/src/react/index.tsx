export { default as ArrowDown } from './ArrowDown';
export { default as ArrowLeft } from './ArrowLeft';
export { default as ArrowRight } from './ArrowRight';
export { default as ArrowUp } from './ArrowUp';
export { default as CloseCircleFill } from './CloseCircleFill';
export { default as CloseCircle } from './CloseCircle';
export { default as Close } from './Close';
export { default as DeleteKey } from './DeleteKey';
export { default as Empty } from './Empty';
export { default as Keyboard } from './Keyboard';
export { default as Minus } from './Minus';
export { default as PlusCircle } from './PlusCircle';
export { default as Plus } from './Plus';
export { default as Search } from './Search';
export { default as SuccessCircle } from './SuccessCircle';
export { default as Success } from './Success';
export { default as Volume } from './Volume';
export { default as WarningCircle } from './WarningCircle';
export { default as Warning } from './Warning';
