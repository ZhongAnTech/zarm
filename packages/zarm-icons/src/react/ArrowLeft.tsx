import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowLeft = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M296.114 508.035c-3.22-13.597.473-28.499 11.079-39.105l333.912-333.912c16.271-16.272 42.653-16.272 58.925 0s16.272 42.654 0 58.926L395.504 498.47l304.574 304.574c16.272 16.272 16.272 42.654 0 58.926s-42.654 16.272-58.926 0L307.241 528.058a41.472 41.472 0 01-11.127-20.023z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowLeft);
export default ForwardRef;
