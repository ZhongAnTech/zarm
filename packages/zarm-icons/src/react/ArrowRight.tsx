import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowRight = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M711.156 508.035a41.472 41.472 0 01-11.126 20.023L366.12 861.97c-16.272 16.272-42.654 16.272-58.926 0s-16.272-42.654 0-58.926L611.767 498.47 307.241 193.944c-16.272-16.272-16.272-42.654 0-58.926 16.271-16.272 42.653-16.272 58.925 0L700.078 468.93c10.606 10.606 14.299 25.508 11.078 39.105z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowRight);
export default ForwardRef;
