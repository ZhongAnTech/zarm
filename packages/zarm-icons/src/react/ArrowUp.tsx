import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgArrowUp = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M513.176 285.05a41.472 41.472 0 0120.024 11.127L867.11 630.089c16.272 16.271 16.272 42.653 0 58.925s-42.654 16.272-58.925 0L503.612 384.44 199.085 688.967c-16.272 16.271-42.654 16.271-58.925 0-16.272-16.272-16.272-42.654 0-58.926L474.07 296.129c10.606-10.606 25.508-14.299 39.105-11.078z"
          fillRule="evenodd"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgArrowUp);
export default ForwardRef;
