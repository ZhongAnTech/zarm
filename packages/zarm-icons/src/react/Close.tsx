import * as React from 'react';
import Icon from '../icon';
import type { IconProps } from '../icon';

const SvgClose = (props: IconProps, svgRef?: React.Ref<SVGSVGElement>) => {
  return React.createElement(Icon, {
    ...props,
    component: () => (
      <svg
        width="1em"
        height="1em"
        viewBox="0 0 1000 1000"
        fill="currentColor"
        focusable={false}
        aria-hidden="true"
        ref={svgRef}
      >
        <path
          d="M560.149 501.223l255.344 255.344c16.272 16.272 16.272 42.654 0 58.926s-42.654 16.272-58.926 0L501.223 560.149 245.88 815.493c-16.272 16.272-42.654 16.272-58.925 0-16.272-16.272-16.272-42.654 0-58.926l255.344-255.344L186.954 245.88c-16.272-16.272-16.272-42.654 0-58.925 16.271-16.272 42.653-16.272 58.925 0l255.344 255.344 255.344-255.344c16.272-16.272 42.654-16.272 58.926 0 16.272 16.271 16.272 42.653 0 58.925L560.149 501.223z"
          fillRule="nonzero"
        />
      </svg>
    ),
  });
};

const ForwardRef = React.forwardRef(SvgClose);
export default ForwardRef;
