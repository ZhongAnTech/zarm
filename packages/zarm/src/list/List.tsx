import * as React from 'react';
import classnames from 'classnames';
import { ConfigContext } from '../n-config-provider';
import ListItem from './ListItem';

export type ListProps = React.HTMLAttributes<HTMLDivElement>;

interface CompoundedComponent
  extends React.ForwardRefExoticComponent<ListProps & React.RefAttributes<HTMLDivElement>> {
  Item: typeof ListItem;
}

const List = React.forwardRef<unknown, ListProps>((props, ref) => {
  const { className, children, ...restProps } = props;

  const compRef = (ref as any) || React.createRef<HTMLDivElement>();

  const { prefixCls: globalPrefixCls } = React.useContext(ConfigContext);
  const prefixCls = `${globalPrefixCls}-list`;
  const cls = classnames(prefixCls, className);

  return (
    <div ref={compRef} className={cls} {...restProps}>
      <ul>{children}</ul>
    </div>
  );
}) as CompoundedComponent;

List.displayName = 'List';

List.defaultProps = {};

export default List;
