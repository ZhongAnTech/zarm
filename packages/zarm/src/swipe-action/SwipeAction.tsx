import { useDrag } from '@use-gesture/react';
import { createBEM } from '@zarm-design/bem';
import React, { useCallback, useContext, useRef } from 'react';
import { ConfigContext } from '../config-provider';
import useClickAway from '../use-click-away';
import { useSafeState } from '../utils/hooks';
import type { HTMLProps } from '../utils/utilityTypes';
import type { BaseSwipeActionItemProps, BaseSwipeActionProps } from './interface';
import SwipeActionItem from './SwipeActionItem';

export interface SwipeActionCssVars {
  '--background'?: React.CSSProperties['background'];
}

export type SwipeActionItemProps = HTMLProps & BaseSwipeActionItemProps;

export type SwipeActionProps = BaseSwipeActionProps &
  React.PropsWithChildren<HTMLProps<SwipeActionCssVars>> & {
    leftActions?: SwipeActionItemProps[];
    rightActions?: SwipeActionItemProps[];
  };

const SwipeAction = React.forwardRef<HTMLDivElement, SwipeActionProps>((props, ref) => {
  const {
    children,
    className,
    leftActions,
    rightActions,
    moveDistanceRatio,
    moveTimeSpan,
    animationDuration: initialAnimationDuration,
    offset,
    autoClose,
    disabled,
    onClose,
    onOpen,
  } = props;

  const isOpen = useRef<null | string>(null);
  const pending = useRef(false);
  const leftRef = useRef<HTMLDivElement>();
  const rightRef = useRef<HTMLDivElement>();
  const swipeActionWrap = useRef<HTMLDivElement | null>((ref as any) || null);
  const { prefixCls } = useContext(ConfigContext);
  const [offsetLeft, setOffsetLeft] = useSafeState<number>(0);
  const [animationDuration, setAnimationDuration] = useSafeState(initialAnimationDuration);
  const bem = createBEM('swipe-action', { prefixCls });

  const doTransition = useCallback(
    ({ offsetX, duration }) => {
      setAnimationDuration(duration);
      setOffsetLeft(offsetX);
    },
    [offsetLeft],
  );

  useClickAway(
    swipeActionWrap.current,
    () => {
      close();
    },
    'ontouchstart' in window ? 'touchstart' : 'mousedown',
  );

  const dragStart = useRef(0);

  const close = () => {
    if (pending.current) return;
    doTransition({ offsetX: 0, duration: initialAnimationDuration });
    isOpen.current = null;
    dragStart.current = 0;
  };

  const renderButtons = (actions, direction, refs) => {
    if (!actions || actions.length === 0) {
      return;
    }

    const cls = bem('actions', [{ [`${direction}`]: true }]);

    return (
      <div className={cls} ref={refs}>
        {actions.map((action, index) => {
          return (
            <SwipeActionItem
              {...action}
              key={+index}
              onClick={async () => {
                pending.current = true;
                await action.onClick?.();
                pending.current = false;
                if (autoClose) {
                  close?.();
                }
              }}
            />
          );
        })}
      </div>
    );
  };

  const dragging = useRef(false);
  const btnsLeftWidth = leftRef?.current?.offsetWidth || 0;
  const btnsRightWidth = rightRef?.current?.offsetWidth || 0;

  const bind = useDrag(
    (state) => {
      const [offsetX] = state.offset;
      if (
        (isOpen.current === 'right' && offsetX < 0) ||
        (isOpen.current === 'left' && offsetX > 0)
      ) {
        return false;
      }
      if (state.down) {
        dragging.current = true;
      }
      if (!dragging.current) return;
      dragStart.current = offsetX;

      if (offsetX > 0 && !leftRef?.current) {
        return false;
      }

      if (offsetX < 0 && !rightRef?.current) {
        return false;
      }

      if (state.last) {
        const timeSpan = Math.floor(state.elapsedTime);
        let distanceX = 0;
        let _isOpen = false;
        if (
          btnsLeftWidth > 0 &&
          (offsetX / btnsLeftWidth > moveDistanceRatio! ||
            (offsetX > 0 && timeSpan <= moveTimeSpan!))
        ) {
          distanceX = btnsLeftWidth;
          _isOpen = true;
        } else if (
          (btnsRightWidth > 0 && offsetX / btnsRightWidth < -moveDistanceRatio!) ||
          (offsetX < 0 && timeSpan <= moveTimeSpan!)
        ) {
          distanceX = -btnsRightWidth;
          _isOpen = true;
        }
        doTransition({ offsetX: distanceX, duration: initialAnimationDuration });

        if (_isOpen) {
          // 打开
          isOpen.current = distanceX > 0 ? 'left' : 'right';
          onOpen?.();
        } else {
          // 还原
          close();
        }
        window.setTimeout(() => {
          dragging.current = false;
        });
      } else {
        doTransition({ offsetX, duration: 0 });
      }
    },
    {
      from: () => [dragStart.current, 0],
      bounds: () => {
        const leftWidth = leftRef?.current?.offsetWidth || 0;
        const rightWidth = rightRef?.current?.offsetWidth || 0;
        return {
          left: -rightWidth - offset!,
          right: leftWidth + offset!,
        };
      },
      enabled: !disabled,
      axis: 'x',
      pointer: { touch: true },
      preventScroll: true,
      triggerAllEvents: true,
    },
  );

  const style = {
    WebkitTransitionDuration: `${animationDuration}ms`,
    transitionDuration: `${animationDuration}ms`,
    WebkitTransform: `translate3d(${offsetLeft}px, 0, 0)`,
    transform: `translate3d(${offsetLeft}px, 0, 0)`,
  };

  const cls = bem([className]);

  return (
    <>
      {leftActions || rightActions ? (
        <div ref={swipeActionWrap} className={cls} style={props.style} {...bind()}>
          {renderButtons(leftActions, 'left', leftRef)}
          {renderButtons(rightActions, 'right', rightRef)}
          <div
            className={bem('content')}
            style={style}
            onTransitionEnd={() => !isOpen.current && onClose?.()}
            onClick={(e) => {
              if (isOpen.current && !dragging.current) {
                e.preventDefault();
                e.stopPropagation();
                close();
              }
            }}
          >
            {children}
          </div>
        </div>
      ) : (
        children
      )}
    </>
  );
});

SwipeAction.displayName = 'SwipeAction';

SwipeAction.defaultProps = {
  leftActions: [],
  rightActions: [],
  moveDistanceRatio: 0.5,
  moveTimeSpan: 300,
  animationDuration: 300,
  offset: 10,
  autoClose: true,
  disabled: false,
};

export default SwipeAction;
