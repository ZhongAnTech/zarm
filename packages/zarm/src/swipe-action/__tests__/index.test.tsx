import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import SwipeAction from '../index';
import List from '../../list/index';

const originalOffsetWidth = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'offsetWidth');

describe('SwipeAction', () => {
  beforeEach(() => {
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', { configurable: true, value: 500 });
  });
  afterAll(() => {
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', originalOffsetWidth!);
  });
  it('renders correctly', () => {
    jest.useFakeTimers();
    const { container } = render(
      <SwipeAction
        leftActions={[
          {
            text: '左按钮1',
            onClick: () => jest.fn(),
          },
          {
            text: '左按钮2',
            theme: 'danger',
            onClick: () => jest.fn(),
          },
        ]}
        rightActions={[
          {
            text: '右按钮1',
            onClick: () => jest.fn(),
          },
          {
            text: '右按钮2',
            theme: 'danger',
            onClick: () => jest.fn(),
          },
        ]}
      >
        <div>左右都能滑动</div>
      </SwipeAction>,
    );
    expect(container).toMatchSnapshot();
    jest.runAllTimers();
  });

  it('left only', () => {
    const { container } = render(
      <SwipeAction
        leftActions={[
          {
            text: '左按钮1',
            onClick: () => jest.fn(),
          },
          {
            text: '左按钮2',
            theme: 'danger',
            onClick: () => jest.fn(),
          },
        ]}
      >
        <div>右滑</div>
      </SwipeAction>,
    );
    expect(container).toMatchSnapshot();
  });

  it('right only', () => {
    const { container } = render(
      <SwipeAction
        rightActions={[
          {
            text: '右按钮1',
            onClick: () => jest.fn(),
          },
          {
            text: '右按钮2',
            onClick: () => jest.fn(),
          },
        ]}
      >
        <div>左滑</div>
      </SwipeAction>,
    );
    expect(container).toMatchSnapshot();
  });

  it('touch event callback', async () => {
    const onOpen = jest.fn();
    const onClose = jest.fn();
    const { container } = render(
      <List>
        <SwipeAction
          onOpen={onOpen}
          onClose={onClose}
          leftActions={[
            {
              text: '左按钮',
              onClick: () => jest.fn(),
            },
          ]}
          rightActions={[
            {
              text: '右按钮',
              onClick: () => jest.fn(),
            },
          ]}
        >
          <List.Item>左右滑看看</List.Item>
        </SwipeAction>
      </List>,
    );
    const element = container.querySelector('.za-swipe-action') as HTMLDivElement;
    fireEvent.mouseDown(element, { pointerId: 1, clientX: 0, clientY: 0, buttons: 1 });
    fireEvent.mouseMove(element, { pointerId: 1, clientX: 200, clientY: 0, buttons: 1 });
    fireEvent.mouseUp(element, { pointerId: 1, clientX: 275 });
    expect(onOpen).toHaveBeenCalled();
    fireEvent.mouseDown(document.body);
    const content = container.querySelector('.za-swipe-action__content');
    fireEvent.transitionEnd(content!);
    await waitFor(() => {
      expect(onClose).toHaveBeenCalled();
    });
    fireEvent.mouseDown(element, { pointerId: 1, clientX: 0, clientY: 0, buttons: 1 });
    fireEvent.mouseMove(element, { pointerId: 1, clientX: -200, clientY: 0, buttons: 1 });
    fireEvent.mouseUp(element, { pointerId: 1, clientX: -275 });
    jest.runAllTimers();
    await waitFor(() => {
      expect(onOpen).toBeCalledTimes(2);
    });
    fireEvent.click(content!);
    fireEvent.transitionEnd(content!);
    await waitFor(() => {
      expect(onClose).toBeCalledTimes(2);
    });
  });

  it('SwipeActionItem click', async () => {
    const onClose = jest.fn();
    const onClick = jest.fn();
    const { container } = render(
      <List>
        <SwipeAction
          autoClose
          onClose={onClose}
          leftActions={[
            {
              text: '左按钮1',
              onClick,
            }
          ]}
        >
          <List.Item>右滑看看</List.Item>
        </SwipeAction>
      </List>,
    );
    const element = container.querySelector('.za-swipe-action') as HTMLDivElement;
    // fireEvent.transitionEnd(container.querySelector('.za-swipe-action__content')!);
    fireEvent.mouseDown(element, { pointerId: 1, clientX: 0, clientY: 0, buttons: 1 });
    fireEvent.mouseMove(element, { pointerId: 1, clientX: 200, clientY: 0, buttons: 1 });
    fireEvent.mouseUp(element, { pointerId: 1, clientX: 275 });
    fireEvent.click(container.querySelector('.za-swipe-action-item__item')!);
    await waitFor(() => {
      expect(onClick).toHaveBeenCalled();
    })
  });
});
