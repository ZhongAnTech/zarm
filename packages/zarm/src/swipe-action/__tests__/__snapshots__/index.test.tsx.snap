// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeAction left only 1`] = `
<div>
  <div
    class="za-swipe-action"
  >
    <div
      class="za-swipe-action__actions za-swipe-action__actions--left"
    >
      <button
        aria-disabled="false"
        class="za-button za-button--primary za-button--md za-button--rect za-swipe-action-item__item"
        type="button"
      >
        <span>
          左按钮1
        </span>
      </button>
      <button
        aria-disabled="false"
        class="za-button za-button--danger za-button--md za-button--rect za-swipe-action-item__item"
        type="button"
      >
        <span>
          左按钮2
        </span>
      </button>
    </div>
    <div
      class="za-swipe-action__content"
      style="transition-duration: 300ms; transform: translate3d(0px, 0, 0);"
    >
      <div>
        右滑
      </div>
    </div>
  </div>
</div>
`;

exports[`SwipeAction renders correctly 1`] = `
<div>
  <div
    class="za-swipe-action"
  >
    <div
      class="za-swipe-action__actions za-swipe-action__actions--left"
    >
      <button
        aria-disabled="false"
        class="za-button za-button--primary za-button--md za-button--rect za-swipe-action-item__item"
        type="button"
      >
        <span>
          左按钮1
        </span>
      </button>
      <button
        aria-disabled="false"
        class="za-button za-button--danger za-button--md za-button--rect za-swipe-action-item__item"
        type="button"
      >
        <span>
          左按钮2
        </span>
      </button>
    </div>
    <div
      class="za-swipe-action__actions za-swipe-action__actions--right"
    >
      <button
        aria-disabled="false"
        class="za-button za-button--primary za-button--md za-button--rect za-swipe-action-item__item"
        type="button"
      >
        <span>
          右按钮1
        </span>
      </button>
      <button
        aria-disabled="false"
        class="za-button za-button--danger za-button--md za-button--rect za-swipe-action-item__item"
        type="button"
      >
        <span>
          右按钮2
        </span>
      </button>
    </div>
    <div
      class="za-swipe-action__content"
      style="transition-duration: 300ms; transform: translate3d(0px, 0, 0);"
    >
      <div>
        左右都能滑动
      </div>
    </div>
  </div>
</div>
`;

exports[`SwipeAction right only 1`] = `
<div>
  <div
    class="za-swipe-action"
  >
    <div
      class="za-swipe-action__actions za-swipe-action__actions--right"
    >
      <button
        aria-disabled="false"
        class="za-button za-button--primary za-button--md za-button--rect za-swipe-action-item__item"
        type="button"
      >
        <span>
          右按钮1
        </span>
      </button>
      <button
        aria-disabled="false"
        class="za-button za-button--primary za-button--md za-button--rect za-swipe-action-item__item"
        type="button"
      >
        <span>
          右按钮2
        </span>
      </button>
    </div>
    <div
      class="za-swipe-action__content"
      style="transition-duration: 300ms; transform: translate3d(0px, 0, 0);"
    >
      <div>
        左滑
      </div>
    </div>
  </div>
</div>
`;
